File: C++/代码随想录/Binary_Tree/Find the attributes of Binary tree/111. Minimum Depth of Binary Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def minDepth(root):
    if not root:
        return 0
    if not root.left and not root.right:
        return 1
    if not root.left:
        return minDepth(root.right) + 1
    if not root.right:
        return minDepth(root.left) + 1
    return min(minDepth(root.left), minDepth(root.right)) + 1

# 测试用例
tree1 = TreeNode(3)
tree1.left = TreeNode(9)
tree1.right = TreeNode(20)
tree1.right.left = TreeNode(15)
tree1.right.right = TreeNode(7)

tree2 = TreeNode(2)
tree2.right = TreeNode(3)
tree2.right.right = TreeNode(4)
tree2.right.right.right = TreeNode(5)
tree2.right.right.right.right = TreeNode(6)

print(minDepth(tree1))  # 2
print(minDepth(tree2))  # 5

File: C++/代码随想录/Binary_Tree/Find the attributes of Binary tree/111. Minimum Depth of Binary Tree.cpp
// 111. Minimum Depth of Binary Tree
// https://leetcode.com/problems/minimum-depth-of-binary-tree/
// clang++ 111.\ Minimum\ Depth\ of\ Binary\ Tree.cpp -std=c++17
#include "../../printUtils.hpp"
#include <climits>

template <typename T>
int minDepth(TreeNode<T>* root) {
    if (root == nullptr) return 0;

    if (root->left == nullptr && root->right == nullptr) return 1;
    
    int leftDepth = root->left ? minDepth(root->left) : INT_MAX;
    int rightDepth = root->right ? minDepth(root->right) : INT_MAX;
    
    return 1 + std::min(leftDepth, rightDepth);
}

int main() {
    // TreeNode<int>* root = buildTree({3, 9, 20, -1, -1, 15, 7}, -1);
    TreeNode<int>* root = buildTree({2, -1, 3, -1, 4, -1, 5, -1, 6}, -1);
    std::cout << "Minimum depth of the binary tree: " << minDepth(root) << std::endl;
    deleteTree(root);
    
    return 0;
}
